Revise the backup and restore guide (#1122)
scalar-labs/scalardb: docs/backup-restore.md

# How to Back Up and Restore Databases Used Through ScalarDB

Since ScalarDB provides transaction capabilities on top of non-transactional or transactional databases non-invasively, you need to take special care to back up and restore the databases in a transactionally consistent way.

This guide describes how to back up and restore the databases that ScalarDB supports.

## Create a backup

How you create a backup depends on which database you're using and whether or not you're using multiple databases. The following decision tree shows which approach you should take.

> ```mermaid

Reviewer: I think we are missing an arrow from `Does the database have transaction support?` to `Perform back up with explicit pausing` for the `No` case.

Author: Good catch, thank you! I've added a `No` arrow for that part.

  A ---->|No| E[Perform <a href='#back-up-with-explicit-pausing'>back up with explicit pausing</a>]
```

### Back up without explicit pausing

If you're using ScalarDB with a single database with support for transactions, you can create a backup of the database even while ScalarDB continues to accept transactions.

One requirement for creating a backup in ScalarDB is that backups for all the ScalarDB-managed tables (including the Coordinator table) need to be transactionally consistent or automatically recoverable to a transactionally consistent state. That means that you need to create a consistent snapshot by dumping all tables in a single transaction.

How to create transactionally consistent backups depends on the type of database. Select your database to see examples for how to create a transactionally consistent backup for ScalarDB.

{% capture notice--warning %}
**Attention**

Before creating a backup, you should consider the safest way to create a transactionally consistent backup of your databases and understand any risks that are associated with the backup process.
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

> <div id="tabset-1">

Reviewer: This doesn't cover all the databases we support. So I think we should note that these are merely examples for some of our supported databases.

Author: Understood. I've added a note to this part.

  <button class="tablinks" onclick="openTab(event, 'MySQL1', 'tabset-1')">MySQL</button>
  <button class="tablinks" onclick="openTab(event, 'PostgreSQL1', 'tabset-1')">PostgreSQL</button>
  <button class="tablinks" onclick="openTab(event, 'SQLite1', 'tabset-1')">SQLite</button>
</div>

<div id="Amazon_RDS_or_Azure_Database_for_MySQL_or_PostgreSQL1" class="tabcontent" markdown="1">

You can restore to any point within the backup retention period by using the automated backup feature.
</div>
<div id="MySQL1" class="tabcontent" markdown="1">

Use the `mysqldump` command with the `--single-transaction` option.
</div>
<div id="PostgreSQL1" class="tabcontent" markdown="1">

Use the `pg_dump` command.
</div>
<div id="SQLite1" class="tabcontent" markdown="1">

Use the `.backup` command with the `.timeout` command as specified in [Special commands to sqlite3 (dot-commands)](https://www.sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_)

For an example, see [BASH: SQLite3 .backup command](https://stackoverflow.com/questions/23164445/bash-sqlite3-backup-command).
</div>
</div>

### Back up with explicit pausing

> Another way to create a transactionally consistent backup is to create a backup while ScalarDB Cluster does not have any outstanding transactions. Creating the backup depends on the following:

Reviewer: ScalarDB Cluster is not included in the community version. We should make a note of that.

Author: After re-reading this part, it's not clear if ScalarDB can do backups with explicit pausing. The original sentence used the wording `ScalarDB cluster`, but does this not mean `ScalarDB Cluster` (the product)? For example, should it be simply `ScalarDB` or something like `a cluster of ScalarDB instances` instead?

Reviewer: The original sentence was quite unclear, but I feel `ScalarDB cluster` was referring to `a cluster of ScalarDB instances`. Regardless, using the word "cluster" in this context seems ambiguous, so I think we should opt for a different term.

Author: Got it, thanks! I've updated the wording here.

Reviewer: @feeblefakie Can you please check this sentence?

By using the [Scalar Admin client tool](https://github.com/scalar-labs/scalar-admin/tree/main/java#scalar-admin-client-tool), you can pause nodes, servers, or applications that implement the Scalar Admin interface without losing ongoing transactions.

{% capture notice--info %}
**Note**

When using a PITR feature, you should minimize the clock drifts between clients and servers by using clock synchronization, such as NTP. Otherwise, the time you get as the paused duration might be too different from the time in which the pause was actually conducted, which could restore the backup to a point where ongoing transactions exist.

In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

How you create a transactionally consistent backup depends on the type of database that you're using. Select your database to see how to create a transactionally consistent backup for ScalarDB.

> <div id="tabset-2">

Reviewer: Ditto. This doesn't cover all the databases we support. So I think we should note that these are merely examples for some of our supported databases.

Author: Understood. I've added a note about that here as well.

</div>
<div id="Cosmos_DB_for_NoSQL2" class="tabcontent" markdown="1">

You must create a Cosmos DB for NoSQL account with a continuous backup policy that has the PITR feature enabled. After enabling the feature, backups are created continuously.

To specify a transactionally consistent restore point, pause your application that is using ScalarDB with Cosmos DB for NoSQL as described in [Back up with explicit pausing](#back-up-with-explicit-pausing).
</div>
<div id="DynamoDB2" class="tabcontent" markdown="1">

You must enable the PITR feature for DynamoDB tables. If you're using [ScalarDB Schema Loader](schema-loader.md) to create schemas, the tool enables the PITR feature for tables by default.

To specify a transactionally consistent restore point, pause your application that is using ScalarDB with DynamoDB as described in [Back up with explicit pausing](#back-up-with-explicit-pausing).
</div>
</div>

## Restore a backup

How you restore a transactionally consistent backup depends on the type of database that you're using. Select your database to see how to restore a transactionally consistent backup for ScalarDB.

> <div id="tabset-3">

Reviewer: Ditto. This doesn't cover all the databases we support. So I think we should note that these are merely examples for some of our supported databases.

Author: Understood. I've added a note about that here as well.

# How to Back Up and Restore Databases Used Through ScalarDB

> Since ScalarDB provides transaction capabilities on top of non-transactional (possibly transactional) databases non-invasively, you need to take special care to back up and restore the databases in a transactionally consistent way.

Author: The phrase `possibly transactional` here and having it in parenthesis is a bit confusing. Can we change this sentence as follows?:

```suggestion
Since ScalarDB provides transaction capabilities on top of non-transactional or transactional databases non-invasively, you need to take special care to back up and restore the databases in a transactionally consistent way.
```

Reviewer 2: Sounds better!

This guide describes how to back up and restore the databases that ScalarDB supports.

## Create a backup

How you create a backup depends on which database you're using and whether or not you're using multiple databases. The following decision tree shows which approach you should take.

```mermaid
flowchart TD
  A[Are you using a single database with ScalarDB?]
  A -->|Yes| B[Does the database have transaction support?]
  B --> C[Perform <a href='#back-up-without-explicit-pausing'>back up without explicit pausing</a>]
  A ---->|No| E[Perform <a href='#back-up-with-explicit-pausing'>back up with explicit pausing</a>]
```

### Back up without explicit pausing

If you're using ScalarDB with a single database with support for transactions, you can create a backup of the database even while ScalarDB continues to accept transactions.

> One requirement for creating a backup in ScalarDB is that backups for all the ScalarDB-managed tables (including the Coordinator table) need to be transactionally consistent or automatically recoverable to a transactionally consistent state. That means that you need to create a consistent snapshot by dumping all tables in a single transaction.

Author: For consistency, can we change `snapshot` to `backup` here?

```suggestion
One requirement for creating a backup in ScalarDB is that backups for all the ScalarDB-managed tables (including the Coordinator table) need to be transactionally consistent or automatically recoverable to a transactionally consistent state. That means that you need to create a consistent backup by dumping all tables in a single transaction.
```

  <button class="tablinks" onclick="openTab(event, 'SQLite1', 'tabset-1')">SQLite</button>
</div>

<div id="Amazon_RDS_or_Azure_Database_for_MySQL_or_PostgreSQL1" class="tabcontent" markdown="1">

You can restore to any point within the backup retention period by using the automated backup feature.
</div>
<div id="MySQL1" class="tabcontent" markdown="1">

Use the `mysqldump` command with the `--single-transaction` option.
</div>
<div id="PostgreSQL1" class="tabcontent" markdown="1">

Use the `pg_dump` command.
</div>
<div id="SQLite1" class="tabcontent" markdown="1">

Use the `.backup` command with the `.timeout` command as specified in [Special commands to sqlite3 (dot-commands)](https://www.sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_)

For an example, see [BASH: SQLite3 .backup command](https://stackoverflow.com/questions/23164445/bash-sqlite3-backup-command).
</div>
</div>

### Back up with explicit pausing

Another way to create a transactionally consistent backup is to create a backup while ScalarDB Cluster does not have any outstanding transactions. Creating the backup depends on the following:

> - If the underlying database has a point-in-time snapshot or backup feature, you can take a snapshot during the period.

Author: For consistency in how we seem to use `backup` more often than snapshot, can we change this sentence as follows?:

```suggestion
- If the underlying database has a point-in-time snapshot or backup feature, you can create a backup during the period.
```

Reviewer 2: Sounds good to me.

- If the underlying database has a point-in-time restore or recovery (PITR) feature, you can set a restore point to a time (preferably the mid-time) in the pause duration period.

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.

Author: Can we change `drain` to `resolve`?

```suggestion
To make ScalarDB resolve outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

Or would `address` be more accurate?:

```suggestion
To make ScalarDB address outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

Reviewer 2: I think `drain` is more appropriate in this case.
(`drain` is a commonly used term for this kind of context in software.)

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.

Author: Should we change `ScalarDB Server` to `ScalarDB Cluster` here?:

```suggestion
To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use ScalarDB Cluster, which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

Reviewer 2: Yes, I think so.

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the `scalar-admin` interface, or implement the [scalar-admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.

Author: Can we change `drain` to `resolve`?

```suggestion
To make ScalarDB resolve outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the `scalar-admin` interface, or implement the [scalar-admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

Or maybe `address` is better?

```suggestion
To make ScalarDB address outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the `scalar-admin` interface, or implement the [scalar-admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

Author: Keeping as is.

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use [ScalarDB Server](scalardb-server.md), which uses the `scalar-admin` interface, or implement the [scalar-admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.

Author: Should we change `ScalarDB Server` to `ScalarDB Cluster` here?

 ```suggestion
To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use ScalarDB Cluster, which uses the `scalar-admin` interface, or implement the [scalar-admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.
```

In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

How you create a transactionally consistent backup depends on the type of database that you're using. Select your database to see how to create a transactionally consistent backup for ScalarDB.

<div id="tabset-2">
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'Cassandra2', 'tabset-2')" id="defaultOpen-2">Cassandra</button>
  <button class="tablinks" onclick="openTab(event, 'Cosmos_DB_for_NoSQL2', 'tabset-2')">Cosmos DB for NoSQL</button>
  <button class="tablinks" onclick="openTab(event, 'DynamoDB2', 'tabset-2')">DynamoDB</button>
</div>

<div id="Cassandra2" class="tabcontent" markdown="1">

> Cassandra has a built-in replication feature, so you do not always have to create a transactionally consistent backup. For example, if the replication factor is set to `3` and only the data of one of the nodes in a Cassandra cluster is lost, you won't need a transactionally consistent backup because the node can be recovered by using a normal, transactionally inconsistent snapshot and the repair feature.

Author: Can we change this sentence as follows since `snapshot` seems to be the phrase used for backups in Cassandra? We use a similar format in the `## Restore a backup` section:

```suggestion
Cassandra has a built-in replication feature, so you do not always have to create a transactionally consistent backup. For example, if the replication factor is set to `3` and only the data of one of the nodes in a Cassandra cluster is lost, you won't need a transactionally consistent backup because the node can be recovered by using a normal, transactionally inconsistent backup (snapshot) and the repair feature.
```

> However, if the quorum of cluster nodes loses their data, you will need a transactionally consistent backup to restore the cluster to a certain transactionally consistent point.

Author: Can we change this sentence as follows since `snapshot` seems to be the phrase used for backups in Cassandra? We use a similar format in the `## Restore a backup` section:

```suggestion
However, if the quorum of cluster nodes loses their data, you will need a transactionally consistent backup (snapshot) to restore the cluster to a certain transactionally consistent point.
```

Reviewer 2: Sounds good to me.

> To create a transactionally consistent cluster-wide backup, pause the application that is using ScalarDB or ScalarDB Server and take snapshots of the nodes as described in [Back up with explicit pausing](#back-up-with-explicit-pausing) or stop the Cassandra cluster, take copies of all the data in the nodes, and start the cluster.

Author: For consistency in how we seem to use `backup` more often than snapshot, can we change this sentence as follows?:

```suggestion
To create a transactionally consistent cluster-wide backup, pause the application that is using ScalarDB or ScalarDB Server and create backups of the nodes as described in [Back up with explicit pausing](#back-up-with-explicit-pausing) or stop the Cassandra cluster, take copies of all the data in the nodes, and start the cluster.
```

> To create a transactionally consistent cluster-wide backup, pause the application that is using ScalarDB or ScalarDB Server and take snapshots of the nodes as described in [Back up with explicit pausing](#back-up-with-explicit-pausing) or stop the Cassandra cluster, take copies of all the data in the nodes, and start the cluster.

Author: Should we change `ScalarDB Server` to `ScalarDB Cluster` here?

```suggestion
To create a transactionally consistent cluster-wide backup, pause the application that is using ScalarDB or ScalarDB Cluster and take snapshots of the nodes as described in [Back up with explicit pausing](#back-up-with-explicit-pausing) or stop the Cassandra cluster, take copies of all the data in the nodes, and start the cluster.
```

Reviewer 2: Yes, I think so.

{% capture notice--info %}
**Note**

> To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy), which is a backup tool for Cassandra. Cassy is also integrated with Scalar Admin, so the tool can issue a pause request to the Cassandra cluster of your application that is using ScalarDB or ScalarDB Server. For more details about how to create a backup by using Cassy, see [Take cluster-wide consistent backups](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups).

Author: Should we change `ScalarDB Server` to `ScalarDB Cluster` here?

```suggestion
To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy), which is a backup tool for Cassandra. Cassy is also integrated with Scalar Admin, so the tool can issue a pause request to the Cassandra cluster of your application that is using ScalarDB or ScalarDB Cluster. For more details about how to create a backup by using Cassy, see [Take cluster-wide consistent backups](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups).
```

Reviewer 3: This is a bit off-topic, but should we archive the GitHub project?

> To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy), which is a backup tool for Cassandra. Cassy is also integrated with Scalar Admin, so the tool can issue a pause request to the Cassandra cluster of your application that is using ScalarDB or ScalarDB Server. For more details about how to create a backup by using Cassy, see [Take cluster-wide consistent backups](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups).

Author: I reworded this sentence, so please confirm its accuracy.

> To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy), which is a backup tool for Cassandra. Cassy is also integrated with `scalar-admin`, so the tool can issue a pause request to the Cassandra cluster of your application that is using ScalarDB or ScalarDB Server. For more details about how to create a backup by using Cassy, see [Take cluster-wide consistent backups](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups).

Author: Should we change `ScalarDB Server` to `ScalarDB Cluster` here?

```suggestion
To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy), which is a backup tool for Cassandra. Cassy is also integrated with `scalar-admin`, so the tool can issue a pause request to the Cassandra cluster of your application that is using ScalarDB or ScalarDB Cluster. For more details about how to create a backup by using Cassy, see [Take cluster-wide consistent backups](https://github.com/scalar-labs/cassy/blob/master/docs/getting-started.md#take-cluster-wide-consistent-backups).
```

  <button class="tablinks" onclick="openTab(event, 'Cassandra3', 'tabset-3')">Cassandra</button>
  <button class="tablinks" onclick="openTab(event, 'Cosmos_DB_for_NoSQL3', 'tabset-3')">Cosmos DB for NoSQL</button>
  <button class="tablinks" onclick="openTab(event, 'DynamoDB3', 'tabset-3')">DynamoDB</button>
  <button class="tablinks" onclick="openTab(event, 'MySQL3', 'tabset-3')">MySQL</button>
  <button class="tablinks" onclick="openTab(event, 'PostgreSQL3', 'tabset-3')">PostgreSQL</button>
  <button class="tablinks" onclick="openTab(event, 'SQLite3', 'tabset-3')">SQLite</button>
</div>

<div id="Amazon_RDS_or_Azure_Database_for_MySQL_or_PostgreSQL3" class="tabcontent" markdown="1">

You can restore to any point within the backup retention period by using the automated backup feature.
</div>
<div id="Cassandra3" class="tabcontent" markdown="1">

First, stop all the nodes of the Cassandra cluster. Then, clean the `data`, `commitlog`, and `hints` directories, and place the backups (snapshots) in each node.

After placing the backups (snapshots) in each node, start all the nodes of the Cassandra Cluster.

{% capture notice--info %}
**Note**

> To avoid potential issues, you should use [Cassy](https://github.com/scalar-labs/cassy). For more details about how to restore a backup by using Cassy, see the following:

Reviewer 2: Can you remove this since `Cassy` is no longer maintained?

</div>

<div id="Amazon_RDS_or_Azure_Database_for_MySQL_or_PostgreSQL1" class="tabcontent" markdown="1">

You can restore to any point within the backup retention period by using the automated backup feature.
</div>
<div id="MySQL1" class="tabcontent" markdown="1">

Use the `mysqldump` command with the `--single-transaction` option.
</div>
<div id="PostgreSQL1" class="tabcontent" markdown="1">

Use the `pg_dump` command.
</div>
<div id="SQLite1" class="tabcontent" markdown="1">

Use the `.backup` command with the `.timeout` command as specified in [Special commands to sqlite3 (dot-commands)](https://www.sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_)

For an example, see [BASH: SQLite3 .backup command](https://stackoverflow.com/questions/23164445/bash-sqlite3-backup-command).
</div>
</div>

### Back up with explicit pausing

Another way to create a transactionally consistent backup is to create a backup while ScalarDB Cluster does not have any outstanding transactions. Creating the backup depends on the following:

- If the underlying database has a point-in-time snapshot or backup feature, you can create a backup during the period.

> - If the underlying database has a point-in-time restore or recovery (PITR) feature, you can set a restore point to a time (preferably the mid-time) in the pause duration period.

Author: That makes sense to me. Thank you for the suggestion!

{% capture notice--info %}
**Note**

When using a PITR feature, you should minimize the clock drifts between clients and servers by using clock synchronization, such as NTP. Otherwise, the time you get as the paused duration might be too different from the time in which the pause was actually conducted, which could restore the backup to a point where ongoing transactions exist.

> In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

Reviewer 2: `5 seconds` sounds good to me.

> In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

Reviewer 4: @feeblefakie 
Thank you for your comment!

@josh-wong 
Could you please update the duration time from 10 seconds to 5 seconds?

> In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

Author: Done! Thank you both for confirming the seconds here!

> In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

Author: Done! Thank you both for confirming the seconds here!

> In addition, you should pause for a sufficient amount of time (for example, 10 seconds) and use the mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes.

Reviewer 4: Thank you for the update!

{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should use ScalarDB Cluster, which uses the Scalar Admin interface, or implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB.

Reviewer: I've just realized ScalarDB Cluster is mentioned here as well. As I pointed out earlier, ScalarDB Cluster is not included in the community version. We should make a note of that.

Author: Got it. I've added notes on both mentions of ScalarDB Cluster.

Reviewer: Thanks! Should we also add a link to the ScalarDB Cluster documentation?

Author: I had considered that, but readers of the Community docs wouldn't be able to access the ScalarDB Cluster documentation since it's only available on the Enterprise docs site.

Also, if we add a static link, we would only be able to add one for the `latest` version (which would be confusing for users who are looking at docs from a previous release).

Reviewer: Understood. We need to discuss how we reference ScalarDB Cluster in the Community docs.

@feeblefakie what are your thoughts on this?

Author: @brfrn169 (cc: @feeblefakie) Regarding the following part, I was incorrect, so let me rephrase:

> Also, if we add a static link, we would only be able to add one for the `latest` version (which would be confusing for users who are looking at docs from a previous release).

We could add static links that point to a ScalarDB Cluster document for this doc on branches `3.9` and `3.10`. 

However, I'm still concerned that Community users might be confused if they click the link and find themselves going from the Community docs site to the Enterprise docs site. If that's a concern we're willing to make an exception for, I'm OK with that.

Reviewer 2: @josh-wong 
Can we do something like this to let users to notice that they are going to the enterprise doc site?

[ScalarDB Cluster (going to Enterprise docs site)](link)

Author: @feeblefakie
I'm OK using "[ScalarDB Cluster (redirects to the Enterprise docs site)](#)". 

@brfrn169 I'll add a link to the page for the version that the user is on when backporting this PR. For this PR, which is for the `master` branch, I'll specify `latest` as the link.

> To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB or use [ScalarDB Cluster (redirects to the Enterprise docs site)](https://scalardb.scalar-labs.com/docs/latest/scalardb-cluster/), which uses the Scalar Admin interface.

Reviewer 2: ```suggestion
To make ScalarDB drain outstanding requests and stop accepting new requests so that a pause duration can be created, you should implement the [Scalar Admin](https://github.com/scalar-labs/scalar-admin) interface properly in your application that uses ScalarDB or use [ScalarDB Cluster (redirects to the Enterprise docs site)](https://scalardb.scalar-labs.com/docs/latest/scalardb-cluster/), which implements the Scalar Admin interface.
```
The Admin interface is something we implement, not use.

> By using the [Scalar Admin client tool](https://github.com/scalar-labs/scalar-admin/tree/main/java#scalar-admin-client-tool), you can pause nodes, servers, or applications that implement the Scalar Admin interface without losing ongoing transactions.

Reviewer 4: FYI:

Now, we have a new tool `scalar-admin-k8s`. So, I think we also should mention it in this document.
https://github.com/scalar-labs/scalar-admin-k8s

However, we still don't have documents for the `scalar-admin-k8s` (i.e., we cannot add a link to any documents for `scalar-admin-k8s`).
I will create those documents, and after that, I will update this document later.

Author: I see. Yes, let's update this document when we have a document to share so that users can reference it👍

</div>
<div id="Cosmos_DB_for_NoSQL2" class="tabcontent" markdown="1">

You must create a Cosmos DB for NoSQL account with a continuous backup policy that has the PITR feature enabled. After enabling the feature, backups are created continuously.

To specify a transactionally consistent restore point, pause your application that is using ScalarDB with Cosmos DB for NoSQL as described in [Back up with explicit pausing](#back-up-with-explicit-pausing).
</div>
<div id="DynamoDB2" class="tabcontent" markdown="1">

You must enable the PITR feature for DynamoDB tables. If you're using [ScalarDB Schema Loader](schema-loader.md) to create schemas, the tool enables the PITR feature for tables by default.

To specify a transactionally consistent restore point, pause your application that is using ScalarDB with DynamoDB as described in [Back up with explicit pausing](#back-up-with-explicit-pausing).
</div>
</div>

## Restore a backup

How you restore a transactionally consistent backup depends on the type of database that you're using. Select a database to see how to create a transactionally consistent backup for ScalarDB.

{% capture notice--warning %}
**Note**

> The backup methods by database listed below are just examples of some of the databases that ScalarDB supports.

Reviewer 4: ```suggestion
The restore methods by database listed below are just examples of some of the databases that ScalarDB supports.
```
It seems that the `restore` is correct here.

Author: Ah, thanks for catching that!

  <button class="tablinks" onclick="openTab(event, 'Cassandra3', 'tabset-3')">Cassandra</button>
  <button class="tablinks" onclick="openTab(event, 'Cosmos_DB_for_NoSQL3', 'tabset-3')">Cosmos DB for NoSQL</button>
  <button class="tablinks" onclick="openTab(event, 'DynamoDB3', 'tabset-3')">DynamoDB</button>
  <button class="tablinks" onclick="openTab(event, 'MySQL3', 'tabset-3')">MySQL</button>
  <button class="tablinks" onclick="openTab(event, 'PostgreSQL3', 'tabset-3')">PostgreSQL</button>
  <button class="tablinks" onclick="openTab(event, 'SQLite3', 'tabset-3')">SQLite</button>
</div>

<div id="Amazon_RDS_or_Azure_Database_for_MySQL_or_PostgreSQL3" class="tabcontent" markdown="1">

You can restore to any point within the backup retention period by using the automated backup feature.
</div>
<div id="Cassandra3" class="tabcontent" markdown="1">

First, stop all the nodes of the Cassandra cluster. Then, clean the `data`, `commitlog`, and `hints` directories, and place the backups (snapshots) in each node.

After placing the backups (snapshots) in each node, start all the nodes of the Cassandra Cluster.
</div>
<div id="Cosmos_DB_for_NoSQL3" class="tabcontent" markdown="1">

> Follow the official Azure documentation for [restore an account by using Azure portal](https://docs.microsoft.com/en-us/azure/cosmos-db/restore-account-continuous-backup#restore-account-portal). After restoring a backup, change the default consistencies of the restored databases to `STRONG`. In addition, you should use the mid-time of the paused duration as the restore point as previously explained.

Reviewer 4: ```suggestion
Follow the official Azure documentation for [restore an account by using Azure portal](https://docs.microsoft.com/en-us/azure/cosmos-db/restore-account-continuous-backup#restore-account-portal). After restoring a backup, [configure the default consistency level](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/how-to-manage-consistency#configure-the-default-consistency-level) of the restored databases to `STRONG`. In addition, you should use the mid-time of the paused duration as the restore point as previously explained.
```
This is a minor thing, but the `default consistency level` are proper word in the official document. So, I think it would be better to use those words instead of `default consistencies`.

Also, I think it would be better to add an official document link to configure the default consistency level of Cosmos DB.

Author: I see. I agree, especially with providing a link👍
